File: frontend/packages/api-client/meals.ts
import { Axios } from 'axios';
import format from 'string-format';

import {
  MealCreationRequestInput,
  Meal,
  MealUpdateRequestInput,
  QueryFilter,
  QueryFilteredResult,
  APIResponse,
} from '@dinnerdonebetter/models';

import { backendRoutes } from './routes';

export async function createMeal(client: Axios, input: MealCreationRequestInput): Promise<Meal> {
  return new Promise(async function (resolve, reject) {
    const response = await client.post<APIResponse<Meal>>(backendRoutes.MEALS, input);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getMeal(client: Axios, mealID: string): Promise<Meal> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<Meal>>(format(backendRoutes.MEAL, mealID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getMeals(
  client: Axios,
  filter: QueryFilter = QueryFilter.Default(),
): Promise<QueryFilteredResult<Meal>> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<Meal[]>>(backendRoutes.MEALS, { params: filter.asRecord() });

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    const result = new QueryFilteredResult<Meal>({
      data: response.data.data,
      totalCount: response.data.pagination?.totalCount,
      page: response.data.pagination?.page,
      limit: response.data.pagination?.limit,
      filteredCount: response.data.pagination?.filteredCount,
    });

    resolve(result);
  });
}

export async function updateMeal(client: Axios, mealID: string, input: MealUpdateRequestInput): Promise<Meal> {
  return new Promise(async function (resolve, reject) {
    const response = await client.put<APIResponse<Meal>>(format(backendRoutes.MEAL, mealID), input);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function deleteMeal(client: Axios, mealID: string): Promise<Meal> {
  return new Promise(async function (resolve, reject) {
    const response = await client.delete<APIResponse<Meal>>(format(backendRoutes.MEAL, mealID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function searchForMeals(client: Axios, query: string): Promise<QueryFilteredResult<Meal>> {
  return new Promise(async function (resolve, reject) {
    const searchURL = `${backendRoutes.MEALS_SEARCH}?q=${encodeURIComponent(query)}`;
    const response = await client.get<APIResponse<Meal[]>>(searchURL);

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    const result = new QueryFilteredResult<Meal>({
      data: response.data.data,
      filteredCount: response.data.pagination?.filteredCount,
      totalCount: response.data.pagination?.totalCount,
      page: response.data.pagination?.page,
      limit: response.data.pagination?.limit,
    });

    resolve(result);
  });
}

File: backend/pkg/apiclient/meals.go
package apiclient

import (
	"context"

	"github.com/dinnerdonebetter/backend/internal/observability"
	"github.com/dinnerdonebetter/backend/internal/observability/keys"
	"github.com/dinnerdonebetter/backend/internal/observability/tracing"
	"github.com/dinnerdonebetter/backend/pkg/apiclient/generated"
	"github.com/dinnerdonebetter/backend/pkg/types"
)

// GetMeal gets a meal.
func (c *Client) GetMeal(ctx context.Context, mealID string) (*types.Meal, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealIDKey, mealID)
	tracing.AttachToSpan(span, keys.MealIDKey, mealID)

	res, err := c.authedGeneratedClient.GetMeal(ctx, mealID)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "get meal")
	}

	var apiResponse *types.APIResponse[*types.Meal]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meal")
	}

	return apiResponse.Data, nil
}

// GetMeals retrieves a list of meals.
func (c *Client) GetMeals(ctx context.Context, filter *types.QueryFilter) (*types.QueryFilteredResult[types.Meal], error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if filter == nil {
		filter = types.DefaultQueryFilter()
	}
	logger = filter.AttachToLogger(logger)
	tracing.AttachQueryFilterToSpan(span, filter)

	params := &generated.GetMealsParams{}
	c.copyType(params, filter)

	res, err := c.authedGeneratedClient.GetMeals(ctx, params)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "meals list")
	}

	var apiResponse *types.APIResponse[[]*types.Meal]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meals")
	}

	response := &types.QueryFilteredResult[types.Meal]{
		Data:       apiResponse.Data,
		Pagination: *apiResponse.Pagination,
	}

	return response, nil
}

// SearchForMeals retrieves a list of meals.
func (c *Client) SearchForMeals(ctx context.Context, query string, filter *types.QueryFilter) (*types.QueryFilteredResult[types.Meal], error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	if filter == nil {
		filter = types.DefaultQueryFilter()
	}

	logger := filter.AttachToLogger(c.logger.Clone())

	tracing.AttachToSpan(span, keys.SearchQueryKey, query)
	tracing.AttachQueryFilterToSpan(span, filter)

	params := &generated.SearchForMealsParams{}
	c.copyType(params, filter)
	params.Q = query

	res, err := c.authedGeneratedClient.SearchForMeals(ctx, params)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "meals list")
	}

	var apiResponse *types.APIResponse[[]*types.Meal]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meals")
	}

	response := &types.QueryFilteredResult[types.Meal]{
		Data:       apiResponse.Data,
		Pagination: *apiResponse.Pagination,
	}

	return response, nil
}

// CreateMeal creates a meal.
func (c *Client) CreateMeal(ctx context.Context, input *types.MealCreationRequestInput) (*types.Meal, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if input == nil {
		return nil, ErrNilInputProvided
	}

	if err := input.ValidateWithContext(ctx); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "validating input")
	}

	body := generated.CreateMealJSONRequestBody{}
	c.copyType(&body, input)

	res, err := c.authedGeneratedClient.CreateMeal(ctx, body)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "create meal")
	}

	var apiResponse *types.APIResponse[*types.Meal]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "creating meal")
	}

	return apiResponse.Data, nil
}

// ArchiveMeal archives a meal.
func (c *Client) ArchiveMeal(ctx context.Context, mealID string) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealIDKey, mealID)
	tracing.AttachToSpan(span, keys.MealIDKey, mealID)

	res, err := c.authedGeneratedClient.ArchiveMeal(ctx, mealID)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "archive meal")
	}

	var apiResponse *types.APIResponse[*types.Meal]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "archiving meal")
	}

	return nil
}
